physical/spanner/spanner_ha.go

```diff
@@ -0,0 +1,409 @@
+package spanner
+
+import (
+	"sync"
+	"time"
+
+	"cloud.google.com/go/spanner"
+	metrics "github.com/armon/go-metrics"
+	uuid "github.com/hashicorp/go-uuid"
+	"github.com/hashicorp/vault/physical"
+	"github.com/pkg/errors"
+	"golang.org/x/net/context"
+	"google.golang.org/grpc/codes"
+)
+
+// Verify Backend satisfies the correct interfaces
+var _ physical.HABackend = (*Backend)(nil)
+var _ physical.Lock = (*Lock)(nil)
+
+const (
+	// LockRenewInterval is the time to wait between lock renewals.
+	LockRenewInterval = 5 * time.Second
+
+	// LockRetryInterval is the amount of time to wait if the lock fails before
+	// trying again.
+	LockRetryInterval = 5 * time.Second
+
+	// LockTTL is the default lock TTL.
+	LockTTL = 15 * time.Second
+
+	// LockWatchRetryInterval is the amount of time to wait if a watch fails
+	// before trying again.
+	LockWatchRetryInterval = 5 * time.Second
+
+	// LockWatchRetryMax is the number of times to retry a failed watch before
+	// signaling that leadership is lost.
+	LockWatchRetryMax = 5
+)
+
+var (
+	// metricLockUnlock is the metric to register for a lock delete.
+	metricLockUnlock = []string{"spanner", "lock", "unlock"}
+
+	// metricLockGet is the metric to register for a lock get.
+	metricLockLock = []string{"spanner", "lock", "lock"}
+
+	// metricLockValue is the metric to register for a lock create/update.
+	metricLockValue = []string{"spanner", "lock", "value"}
+)
+
+// Lock is the HA lock.
+type Lock struct {
+	// backend is the underlying physical backend.
+	backend *Backend
+
+	// key is the name of the key. value is the value of the key.
+	key, value string
+
+	// held is a boolean indicating if the lock is currently held.
+	held bool
+
+	// identity is the internal identity of this key (unique to this server
+	// instance).
+	identity string
+
+	// lock is an internal lock
+	lock sync.Mutex
+
+	// stopCh is the channel that stops all operations. It may be closed in the
+	// event of a leader loss or graceful shutdown. stopped is a boolean
+	// indicating if we are stopped - it exists to prevent double closing the
+	// channel. stopLock is a mutex around the locks.
+	stopCh   chan struct{}
+	stopped  bool
```

Comment:
Rather than a bool and lock you may want to just use atomic.CompareAndSwap

Reply:
Hmm - with channels? Sorry - I've never used CompareAndSwap, so I'm not sure what you're suggesting here.

Reply:
In place of using `stopped`/`stopLock`. You can have a uint32 that defaults to zero; when stopping you atomic.CompareAndSwap and if it was swapped, you continue on, and if not, you're already running stop logic. Then in Lock you do the opposite (compare to 1 and swap to 0).

Reply:
This isn't a must btw, just a suggestion.

Reply:
Gotcha. I would prefer to leave it as-is, if that's okay with you.